quinn/src/endpoint.rs

```diff
@@ -300,15 +299,98 @@ impl Future for EndpointDriver {
 
     #[allow(unused_mut)] // MSRV
     fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
-        let mut endpoint = self.0.lock().unwrap();
+        let mut endpoint = &mut *self.0.lock().unwrap();
         if endpoint.driver.is_none() {
             endpoint.driver = Some(cx.waker().clone());
         }
 
-        let now = Instant::now();
-        let mut keep_going = false;
-        keep_going |= endpoint.drive_recv(cx, now)?;
-        keep_going |= endpoint.handle_events(cx);
+        let mut keep_going = endpoint.drive_recv(cx, Instant::now())?;
+
+        while let Some(conn_handle) = endpoint
+            .timers
+            .poll((Instant::now() - endpoint.timer_epoch).as_millis() as u64)
+        {
+            let conn = match endpoint.connections.connections.get(&conn_handle) {
+                Some(c) => c,
+                None => continue,
+            };
+            let mut conn = &mut *conn.lock("poll timeouts");
+            let _guard = conn.span.clone().entered();
+            conn.inner.handle_timeout(Instant::now());
+            conn.timer_handle = None;
+            conn.timer_deadline = None;
+            conn.wake();
+        }
+        if let Some(deadline) = endpoint.timers.next_timeout() {
+            let deadline = endpoint.timer_epoch + std::time::Duration::from_millis(deadline);
+            let timer = match endpoint.base_timer {
+                Some(ref mut x) => {
+                    x.as_mut().reset(deadline);
+                    x
+                }
+                None => endpoint
+                    .base_timer
+                    .insert(endpoint.runtime.new_timer(deadline)),
+            };
+            if let Poll::Ready(()) = timer.as_mut().poll(cx) {
+                keep_going = true;
+            }
+        }
+
+        let mut dirty = Vec::new();
```

Comment:
could be a member of the endpoint to avoid allocating it in every iteration. Same for `drained`, but that one likely matters less since connection disposure is rare.

Reply:
Done.